File: blockchain/02-OutputsInputsTransactions/src/main/java/no/kalli/Transaction.java
package no.kalli;

import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;

public class Transaction {

    //Simplified compared to Bitcoin
    private List<Input> inputs = new ArrayList<>();
    private List<Output> outputs = new ArrayList<>();

    //If we make the assumption that all the inputs belong to the
    //same key, we can have one signature for the entire transaction,
    //and not one for each input. This simplifies things a lot
    //(more than you think)!
    private PublicKey senderPublicKey;
    private byte[] signature;

    private String txHash;

    public Transaction(PublicKey senderPublicKey) {
        //TODO
    }

    public void addInput(Input input) {
        //TODO
    }

    public void addOutput(Output output) {
        //TODO
    }

    @Override
    public String toString() {
        //TODO
        return null;
    }

    public void signTxUsing(PrivateKey privateKey) {
        //TODO
    }

    public void calculateTxHash() {
        //TODO
    }

    public boolean isValid() {
        //TODO Complete validation of the transaction. Called by the Application.
        return true;
    }

    //TODO Getters?

}

File: blockchain/02-OutputsInputsTransactions/src/main/kotlin/no/kalli/Transaction.kt
package no.kalli

import java.security.PrivateKey
import java.security.PublicKey
import java.util.*
import java.util.stream.Collectors

class Transaction(var senderPublicKey: PublicKey) {

    val inputs = ArrayList<Input>()
    val outputs = ArrayList<Output>()

    private var signature: ByteArray = byteArrayOf(0)

    var txHash: String = ""

    fun isValid(): Boolean {
        return inputs.isNotEmpty()
                && outputs.isNotEmpty()
                && signature.isNotEmpty()
                && outputs.stream().allMatch { it.value < 21000000 && it.value > 0 }
                && DSAUtil.verifyWithDSA(senderPublicKey, inputsToString() + outputsToString(), signature)
                && HashUtil.base64Encode(HashUtil.sha256Hash(inputsToString() + outputsToString() + signature)) == txHash
    }

    fun addInput(input: Input) =
            if (!inputs.contains(input)) inputs.add(input)
            else throw Exception("Input already exists in the lists")

    fun addOutput(output: Output) = outputs.add(output)

    override fun toString(): String =
            "Trasaction( $txHash )\n\tInputs:\n\t\t${inputsToString()}\n\tOutputs:\n\t\t${outputsToString()}"

    fun signTxUsing(privateKey: PrivateKey) {
        signature = DSAUtil.signWithDSA(
                privateKey, inputsToString() + outputsToString())
    }

    fun calculateTxHash() {
        txHash = HashUtil.base64Encode(
                HashUtil.sha256Hash(inputsToString() + outputsToString() + signature))
    }

    private fun inputsToString(): String = inputs.stream().map(Input::toString).collect(Collectors.joining("\n\t\t"))

    private fun outputsToString(): String = outputs.stream().map(Output::toString).collect(Collectors.joining("\n\t\t"))

}